File: DSA/Pranvkumar_590011587/Day51_Question3.py
def min_window_substring(s, t):
    from collections import Counter
    if not t or not s:
        return ""
    t_freq = Counter(t)
    window = {}
    left = 0
    min_len = float('inf')
    res = ""
    have, need = 0, len(t_freq)
    for right, char in enumerate(s):
        window[char] = window.get(char, 0) + 1
        if char in t_freq and window[char] == t_freq[char]:
            have += 1
        while have == need:
            if right - left + 1 < min_len:
                min_len = right - left + 1
                res = s[left:right+1]
            window[s[left]] -= 1
            if s[left] in t_freq and window[s[left]] < t_freq[s[left]]:
                have -= 1
            left += 1
    return res

s = "ADOBECODEBANC"
t = "ABC"
result = min_window_substring(s, t)
print(result) 

File: DSA/pranavdixit_590014418/Day_51_Question_3.c
#include <stdio.h>
#include <string.h>
#include <limits.h>

char* minWindow(char* s, char* t) {
    static char ans[1000];
    int need[256] = {0}, have[256] = {0};
    int required = 0, formed = 0;
    for (int i = 0; t[i]; i++) {
        if (need[t[i]] == 0) required++;
        need[t[i]]++;
    }
    int left = 0, minLen = INT_MAX, start = 0;
    for (int right = 0; s[right]; right++) {
        have[s[right]]++;
        if (have[s[right]] == need[s[right]]) formed++;
        while (formed == required) {
            if (right - left + 1 < minLen) {
                minLen = right - left + 1;
                start = left;
            }
            have[s[left]]--;
            if (have[s[left]] < need[s[left]]) formed--;
            left++;
        }
    }
    if (minLen == INT_MAX) return "";
    strncpy(ans, s + start, minLen);
    ans[minLen] = '\0';
    return ans;
}

int main() {
    char s[] = "ADOBECODEBANC";
    char t[] = "ABC";
    printf("%s", minWindow(s, t));
    return 0;
}
